source/lib-intro.tex

```diff
@@ -1891,9 +1891,9 @@
   \tcode{X::const_pointer}                &
   \tcode{static_cast<} \tcode{X::const_pointer\brk{}>(x) == q} & \\ \rowsep
 
-\tcode{pointer_traits<\brk{}X::pointer\brk{}>::pointer_to(r)}  &
+\tcode{pointer_traits<\brk{}X::pointer\brk{}>::pointer_to(*p)}  &
   \tcode{X::pointer}        &
-                            & \\ \rowsep
+  same as \tcode{p}         & \\ \rowsep
```

Comment:
This looks correct, but is this a novel normative requirement? (Or does this follow from existing wording somehow?)

Reply:
`If `pointer_to` is not the inverse of `*p` then `pointer_to` is unusable for the purpose it was designed for. Formally, `pointer_to(*p)` returns a pointer to `*p`, and I can't imagine a sane pointer-like type where that pointer isn't equal to `p`. Maybe it should be stated in terms of equality, rather than "same as".

Or we could simply define `r` before the table.

Reply:
It sounds like you're saying this is a novel normative requirement, but is one that was intended to be required and omitted, right? If you say this does not need LWG review, I'll take your word for that, but my inclination would be to run this past LWG.

Reply:
I'll take it to LWG.

Reply:
I raised this on the lib reflector and @CaseyCarter cleverly pointed out:

> Per [allocator.requirements]/5, an allocator's pointer types must meet the random access iterator requirements: "`X​::​pointer` and `X​::​const_­pointer` shall also satisfy the requirements for a random access iterator and of a contiguous iterator." Random access iterators must satisfy the forward iterator requirements. Per [forward.iterators]/6, two forward iterators that denote the same object must be equal: "If `a` and `b` are both dereferenceable, then `a == b` if and only if `*a` and `*b` are bound to the same object." Per [pointer.traits.functions]/2, `pointer_to` returns a pointer that denotes its argument: "The first member function returns a pointer to `r` obtained by calling `Ptr​::​pointer_­to(r)` through which indirection is valid; ... ".

> So: given a fancy pointer `p` of type `T` that denotes an object `o`, `std::pointer_traits<T>::pointer_to(*p) == p`.

> The trail is a bit subtle. Perhaps a note would help?

So it's not a novel requirement, it's already there if you look hard enough. The part I was missing was "`a == b` if and only if `*a` and `*b` are bound to the same object" which forbids fancy pointers from having other "salient attributes" that determine equality, only the identity of the pointee determines equality.

Reply:
OK, if this is not a novel requirement, it sounds like explicitly listing it here would improve clarity without changing intent. Looks good.